pyrokinetics/local_geometry/LocalGeometryFourier.py

```diff
@@ -0,0 +1,662 @@
+        # bunit_over_b0 should be an optional argument, and the following should
+        # be performed within __init__ if it is None
+        fourier = cls(params)
+        fourier.bunit_over_b0 = fourier.get_bunit_over_b0()
+        return fourier
+
+    @classmethod
+    def from_global_eq(cls, global_eq: Equilibrium, psi_n: float, verbose=False):
+        # TODO this should replace load_from_eq.
+        fourier = cls()
+        fourier.load_from_eq(global_eq, psi_n=psi_n, verbose=verbose)
+        return fourier
+
+    def load_from_eq(self, eq: Equilibrium, psi_n: float, verbose=False):
+        r"""
+        Loads fourier object from a GlobalEquilibrium Object
+
+        Flux surface contours are fitted from 2D psi grid
+        Gradients in shaping parameters are fitted from poloidal field
+
+        Parameters
+        ----------
+        eq : GlobalEquilibrium
+            GlobalEquilibrium object
+        psi_n : Float
+            Value of :math:`\psi_N` to generate local fourier parameters
+        verbose : Boolean
+            Controls verbosity
+
+        """
+
+        R, Z = eq.get_flux_surface(psi_n=psi_n)
+
+        b_poloidal = eq.get_b_poloidal(R, Z)
+
+        R_major = eq.R_major(psi_n)
+
+        rho = eq.rho(psi_n)
+
+        r_minor = rho * eq.a_minor
+
+        kappa = (max(Z) - min(Z)) / (2 * r_minor)
+
+        Zmid = (max(Z) + min(Z)) / 2
+
+        Zind = np.argmax(abs(Z))
+
+        R_upper = R[Zind]
+
+        fpsi = eq.f_psi(psi_n)
+        B0 = fpsi / R_major
+
+        drho_dpsi = eq.rho.derivative()(psi_n)
+        shift = eq.R_major.derivative()(psi_n) / drho_dpsi / eq.a_minor
+        dZ0dr = eq.Z_mid.derivative()(psi_n) / drho_dpsi / eq.a_minor
+
+        pressure = eq.pressure(psi_n)
+        q = eq.q(psi_n)
+
+        dp_dpsi = eq.q.derivative()(psi_n)
+
+        shat = rho / q * dp_dpsi / drho_dpsi
+
+        dpressure_drho = eq.p_prime(psi_n) / drho_dpsi
+
+        beta_prime = 8 * pi * 1e-7 * dpressure_drho / B0**2
+
+        normalised_height = (Z - Zmid) / (kappa * r_minor)
+
+        # Floating point error can lead to >|1.0|
+        normalised_height = np.where(
+            np.isclose(normalised_height, 1.0), 1.0, normalised_height
+        )
+        normalised_height = np.where(
+            np.isclose(normalised_height, -1.0), -1.0, normalised_height
+        )
+
+        theta = np.arcsin(normalised_height)
+
+        normalised_radius = (R - R_major) / r_minor
+
+        normalised_radius = np.where(
+            np.isclose(normalised_radius, 1.0), 1.0, normalised_radius
+        )
+        normalised_radius = np.where(
+            np.isclose(normalised_radius, -1.0), -1.0, normalised_radius
+        )
+
+        thetaR = np.arccos(normalised_radius)
+
+        for i in range(len(theta)):
+            if R[i] < R_upper:
+                theta[i] = np.pi - theta[i]
+            else:
+                if Z[i] < 0:
+                    theta[i] = 2 * np.pi + theta[i]
```

Comment:
We can get rid of a level of indentation:

```suggestion
            elif Z[i] < 0:
                 theta[i] = 2 * np.pi + theta[i]
```

The following is a bit shorter (and plausibly faster), but not 100% equivalent:

```suggestion
        theta = np.where(R < R_upper, np.pi - theta, theta)
        theta = np.where(Z < 0, 2 * np.pi + theta, theta)
```

Reply:
We need theta to go from `0 -> 2 pi`, but `arcos` only return values between `0` and `pi` 

So we need to rotate some of the angles around depending on where it is on the flux surface. Here for example we need to rotate 3 of the 4 quadrants of a flux surface, firstly when
`R[i] < R_upper` which is the inboard side  (left quadrants)

and when both `R[i] > R_upper` and `Z[i] < 0.0` (bottom right quadrant)

The top right quadrant is already at the correct angle

Reply:
Used double conditional in where to do this